## CMSC 678 HOMEWORK 2 (Q2)
### VAISHNAVI JAMDADE(TM39453)

In [1]:
import numpy as np
import pandas as pd
from numpy import loadtxt
from sklearn.metrics import accuracy_score
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from urllib.request import urlopen

### Reading input data : 

In [2]:
url1="https://www.csee.umbc.edu/courses/graduate/678/spring20/mnist_data.txt"
url2="https://www.csee.umbc.edu/courses/graduate/678/spring20/mnist_labels.txt"

In [3]:
raw_data = urlopen(url1)
dataset1 = loadtxt(raw_data, delimiter=" ")
print(dataset1.shape)
dataset1=dataset1.astype(int)

(10000, 784)


In [4]:
raw_data2 = urlopen(url2)
dataset2 = loadtxt(raw_data2, delimiter=" ")
dataset2=dataset2.astype(int)
dataset2

array([7, 2, 1, ..., 4, 5, 6])

### Function to prepend 1 to input data for bias term :

In [5]:
def add_ones(X):
    
    return np.hstack((np.ones((X.shape[0], 1),dtype=int), X)) 

In [6]:
dataset1=add_ones(dataset1)
dataset1

array([[1, 0, 0, ..., 0, 0, 0],
       [1, 0, 0, ..., 0, 0, 0],
       [1, 0, 0, ..., 0, 0, 0],
       ...,
       [1, 0, 0, ..., 0, 0, 0],
       [1, 0, 0, ..., 0, 0, 0],
       [1, 0, 0, ..., 0, 0, 0]])

### Splitting training and testing data : 

In [7]:
X_train, X_test, y_train, y_test = train_test_split(dataset1,dataset2, test_size=0.5, random_state=42)

In [8]:
X_train.shape,y_train.shape,X_test.shape,y_test.shape

((5000, 785), (5000,), (5000, 785), (5000,))

## Training Function : 

In [9]:

    def training(X_train, y_train):
        
        dotprod=np.ndarray(10,dtype=np.float32)
        weights=np.zeros((10,X_train.shape[1]), dtype=np.float32)
        iterations=47                           # Number of epochs
        alpha=0.01                              # Learning Rate
        accu=[]
        co=[]
        for iter in range(iterations) : 
            count=0
            for i in range(X_train.shape[0]): # iterating through each feature vector
                for j in range(10):     # weight vectors equivalent to 10 class labels 
                    dotprod[j] = np.dot(X_train[i],weights[j])
        
                ind=y_train[i]                 # true class label
                maxi=np.argmax(dotprod,axis=None)
                y_pred=maxi             # got predicted corresponding class label
       
                if(y_pred!=y_train[i]):
                #subtracting x input from old weight vector coress to predicted label 
                    weights[maxi] -= alpha*X_train[i]
                  
                #adding x input to old weight vector coress to true label 
                    weights[ind] += alpha*X_train[i]
                    
                elif(y_pred==y_train[i]):
                    count=count+1
                    
            accuracy = round(((count/y_train.shape[0])*100), 2)
            #print("Accuracy after epoch "+ str(iter)  + " is : ",accuracy)
            accu.append(accuracy)
            #print("Count of correctly classified samples is : " , count)
            co.append(count)
            
        zippedList =  list(zip(accu,co))
        df=pd.DataFrame(zippedList, columns = ['Accuracy in %', 'Correctly_classified_samples'],index=None)
        print(df)
 
        return weights

## Testing Function : 

In [10]:
    def testing(X_test,w):
        
        ypred=[]
        activ=[]
        activ=np.dot(X_test,w.T)     #calculating activation
        
        for i in range(activ.shape[0]):
            ypred.append(np.argmax(activ[i]))

        return ypred           #predicted class label

In [11]:
 w=training(X_train, y_train)  #got updated weights

    Accuracy in %  Correctly_classified_samples
0           77.78                          3889
1           85.62                          4281
2           87.40                          4370
3           87.84                          4392
4           89.48                          4474
5           89.82                          4491
6           90.42                          4521
7           90.54                          4527
8           91.64                          4582
9           91.90                          4595
10          92.04                          4602
11          92.76                          4638
12          92.86                          4643
13          93.34                          4667
14          92.92                          4646
15          93.06                          4653
16          93.56                          4678
17          94.40                          4720
18          94.16                          4708
19          93.90                       

## Accuracy Calculation for Training DATA

In [12]:
predictions_for_Training=testing(X_train,w)         

In [13]:
count=0
for i in range(y_train.shape[0]):
    if predictions_for_Training[i]==y_train[i]:
        count=count+1
print("Number of samples correctly classified for training data : ", count)
print('Accuracy in % :', accuracy_score(y_train,predictions_for_Training)*100)

Number of samples correctly classified for training data :  4902
Accuracy in % : 98.04


## Accuracy Calculation for Testing DATA

In [14]:
predictions_for_Testing=testing(X_test,w)

In [15]:
count=0
for i in range(y_test.shape[0]):
    if predictions_for_Testing[i]==y_test[i]:
        count=count+1

print("Number of samples correctly classified for testing data : ", count)
print('Accuracy in % :', accuracy_score(y_test,predictions_for_Testing)*100)

Number of samples correctly classified for testing data :  4431
Accuracy in % : 88.62


## Code Explanation :

* Here I have implemented a single layer, Multi-class perceptron. I have prepended a column of 1's to the input data for simplification in case of bias term. Further, data has been split 50-50% of training and testing data. A numpy weight array of dimensions(10,785) where (785 since 784 dimensions + bias term) has been declared.
* In the training function, our model will learn based on our training data. With learning rate of 0.01 at each epoch, it calculates the dot product of each feature vector in input data with all the 10 weight vectors coressponding to the given 10 class labels. The index of the maximum dot product-value will be our coressponding  predicted class label.
* If our predictions are correct, then simply store the count of correct prediction at each iteration. But if our prediction is incorrect, then the weights for the correct class are increased by x and the weights for the incorrectly predicted class are decreased by x. 

* Keep updating the weights for each iterations/epoch and compute the training accuracy to see variations. We will see that the accuracy has increased gradually. But after a certain epoch, we may see that accuracy tends to be almost 100% that is tends to overfit our model on training data and perform poorly on testing data.

* Testing function simply takes these updated weights retrieved from training function and runs our model on test data thus returning predicted class labels. 
* Further, accuracy is calculated for test data. I have tuned my weight parameters for 47 epochs and then tested my model since after 48th epoch, it was observed that the although the training accuracy was substantially increasing, the testing accuracy began to decline noticably, that is, it was trying to overfit the model. 
